pkg/k8s/config_map_cache_informer.go

```diff
@@ -0,0 +1,141 @@
+package k8s
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/informers"
+	infcorev1 "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+)
+
+type InformerConfigMap struct {
+	lggr       logr.Logger
+	cmInformer infcorev1.ConfigMapInformer
+	bcaster    *watch.Broadcaster
+	cmObjCh    chan *corev1.ConfigMap
+}
+
+func (i *InformerConfigMap) MarshalJSON() ([]byte, error) {
+	lst := i.cmInformer.Lister()
+	cms, err := lst.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	return json.Marshal(&cms)
+}
+
+func (i *InformerConfigMap) Start(ctx context.Context) error {
+	i.cmInformer.Informer().Run(ctx.Done())
+	return errors.Wrap(
+		ctx.Err(),
+		"configMap informer was stopped",
+	)
+}
+
+func (i *InformerConfigMap) Get(
+	ns,
+	name string,
+) (corev1.ConfigMap, error) {
+	cm, err := i.cmInformer.Lister().ConfigMaps(ns).Get(name)
+	if err != nil {
+		return corev1.ConfigMap{}, err
+	}
+	return *cm, nil
+}
+
+func (i *InformerConfigMap) Watch(
+	ns,
+	name string,
+) watch.Interface {
+	return watch.Filter(i.bcaster.Watch(), func(e watch.Event) (watch.Event, bool) {
+		cm, ok := e.Object.(*corev1.ConfigMap)
+		if !ok {
+			i.lggr.Error(
+				fmt.Errorf("informer expected ConfigMap, ignoring this event"),
+				"event",
+				e,
+			)
+			return e, false
+		}
+		if cm.Namespace == ns && cm.Name == name {
+			return e, true
+		}
+		return e, false
+	})
+}
+
+func (i *InformerConfigMap) addEvtHandler(obj interface{}) {
+	cm, ok := obj.(*corev1.ConfigMap)
+	if !ok {
+		i.lggr.Error(
+			fmt.Errorf("informer expected configMap, got %v", obj),
+			"not forwarding event",
+		)
+		return
+	}
+	i.cmObjCh <- cm
+	i.bcaster.Action(watch.Added, cm)
+}
+
+func (i *InformerConfigMap) updateEvtHandler(oldObj, newObj interface{}) {
+	cm, ok := newObj.(*corev1.ConfigMap)
+	if !ok {
+		i.lggr.Error(
+			fmt.Errorf("informer expected configMap, got %v", newObj),
+			"not forwarding event",
+		)
+		return
+	}
+	i.cmObjCh <- cm
+	i.bcaster.Action(watch.Modified, cm)
```

Comment:
can we unify these two actions, and have the goroutine listening on this channel just listen on a broadcaster channel?

Reply:
As it should be. The reason why a cmObjCh channel was added was that the code I wrote was causing the broadcaster to block, now that I've fixed the broadcaster blocking problem, everything will communicate through the broadcaster.